plutus-ir/src/Language/PlutusIR/Transform/LetFloat.hs

```diff
@@ -0,0 +1,504 @@
+                                                                                   , if isSingleton scc then oRec else Rec
+                                                                                     -- lift the binding into a semigroup for accumulation
+                                                                                   , pure oBinding)
+                                                                    ) res
+              pure $ case accTerm of
+                       -- MERGE current let-group with previous let-group iff both groups' recursivity is NonRec
+                       Let accAnn NonRec accBs accIn | mRecurs == NonRec -> Let (mAnn <> accAnn) NonRec (newBindings <> accBs) accIn
+                       -- never merge if the previous let-group is a Rec or the current let-group is Rec,
+                       -- but instead create a nested current let-group under the previous let-group (above)
+                       _ -> Let mAnn mRecurs newBindings accTerm
+
+
+-- | The main transformation function (Term -> Term) that returns a term with all lets floated as outwards as possible (full laziness).
+-- Is comprised of two AST "passes":
+-- 1stpass: to collect the ranks (positions) of all lets
+-- 2ndpass:  to remove all its lets and place them back (float them) in their ranks (positions).
+-- See Note [Float algorithm]
+--
+-- NB: This transformation requires that the PLC.rename compiler-pass has prior been run.
+floatTerm :: forall name tyname uni a
+           . (PLC.HasUnique (tyname a) PLC.TypeUnique, PLC.HasUnique (name a) PLC.TermUnique, Monoid a)
+          => Term tyname name uni a -> Term tyname name uni a
+floatTerm pir = p2Term pir
+                -- give the floatdata of the 1st pass to the start of the 2nd pass
+              $ p1Term pir
+
+
+
+-- Helpers
+----------
+
+-- | A getter that returns a single 'Unique' for a particular binding.
+-- We need this because let-datatypes introduce multiple identifiers, but in our 'RhsTable', we use a single Unique as the key. See also: 'bindingIds'
+principal :: (PLC.HasUnique (tyname a) PLC.TypeUnique, PLC.HasUnique (name a) PLC.TermUnique)
+            => Getting r (Binding tyname name uni a) PLC.Unique
+principal = to $ \case TermBind _ _ (VarDecl _ n _) _ -> n ^. PLC.theUnique
+                       TypeBind _ (TyVarDecl _ n _) _ -> n ^. PLC.theUnique
+                       -- arbitrary: uses the type construtors' unique as the principal unique of this data binding group
+                       DatatypeBind _ (Datatype _ (TyVarDecl _ tConstr _) _ _ _) -> tConstr ^. PLC.theUnique
+
+
+-- | given the lets to float here, it splits the SCC state to a pair of:
+-- 1) sccs to process at this location
+-- 2) and remainder sccs to use as the next state.
+splitSccs :: Ord a
+          => S.Set a -> [S.Set a] -> ([S.Set a], [S.Set a])
+splitSccs lets =
+  let initAcc = (lets, (mempty,mempty))
+  in snd . foldr (\ scc (ls, (hs,rs)) ->
```

Comment:
Passing a state containing remaining lets from right to left feels weird to me. Not a wrong thing to do I guess, but worth a comment at least.

Reply:
I think the order does not really matter here: the SCCs do not share uniques between each other, so the splitting done from left-to-right or right-to-left would return the same results.

Luck you :) , I changed this to a manual left-to-right visit which stops a bit earlier the splitting when there are no remaining lets to look for. I am a bit doubting though for the complexity of the old as well as the new `splitSccs` function. Can somebody take a look at the new `splitSccs`? 

```haskell
-- | given the lets to float here, it splits the SCC state to a pair of:
-- 1) sccs to process at this location
-- 2) and remainder sccs to use as the next state.
splitSccs :: Ord a
          => NS.NESet a -> [NS.NESet a] -> ([NS.NESet a], [NS.NESet a])
splitSccs _ [] = ([],[]) -- reached end of sccs, done splitting
splitSccs ls (scc:sccs) =
   let commonScc = scc `NS.intersection` ls
       remLets = NS.toSet ls S.\\ commonScc
       remScc = NS.toSet scc S.\\ commonScc
       (hres,rres) = case remLets of
                       NS.IsEmpty             -> ([],sccs) -- early done of splitting
                       NS.IsNonEmpty nremLets -> splitSccs nremLets sccs
   in (commonScc ?: hres, remScc ?: rres)
  where
    -- append a set to a list of nonsets iff it is nonempty
    (?:) :: S.Set a -> [NS.NESet a] -> [NS.NESet a]
    NS.IsNonEmpty n ?: l = n:l
    _ ?: l = l -- id otherwise
```

Reply:
The split stops a bite earlier; I am just worried that the stack is growing (similar to what happens with `foldl`).

Reply:
It's unlikely that you're going to blow up the stack this way, but if you're worried, you can transform the algorithm into a tail-recursive one by collecting `hres`s in a list (and `rres`s in their own) and then folding that lsit with `?:` in the base case of recursion.

Check [this blog post](https://www.joachim-breitner.de/blog/620-Constructing_a_list_in_a_Monad) out.

Another option is to simply attach `~` to `(hres,rres)`. This way the function will become productive and the consumer will force each step of recursion (instead of waiting for recursion to finish to be able to consume the produced data). You can try creating a similar, but simpler function that has the same pattern, and checking if it doesn't consume stack indeed.